Wikipedia/Code/Article as a Living Document/ArticleAsLivingDocViewController.swift

```diff
@@ -162,19 +162,67 @@ class ArticleAsLivingDocViewController: ColumnarCollectionViewController {
     }
 
     func appendSections(_ sections: [ArticleAsLivingDocViewModel.SectionHeader]) {
+		guard let dayMonthNumberYearDateFormatter = DateFormatter.wmf_monthNameDayOfMonthNumberYear(), let isoDateFormatter = DateFormatter.wmf_iso8601() else {
+			return
+		}
 
         var currentSnapshot = dataSource.snapshot()
-
         var existingSections: [ArticleAsLivingDocViewModel.SectionHeader] = []
+
         for currentSection in currentSnapshot.sectionIdentifiers {
             for proposedSection in sections {
                 if currentSection == proposedSection {
-                    currentSnapshot.appendItems(proposedSection.typedEvents, toSection: currentSection)
-                    existingSections.append(proposedSection)
+                    if
+                        let lastCurrentEvent = currentSection.typedEvents.last, lastCurrentEvent.isSmall,
+                        let firstProposedEvent = proposedSection.typedEvents.first, firstProposedEvent.isSmall {
+                        // Collapse sequential small events if appending to the same section
+                        let smallChanges = lastCurrentEvent.smallChanges + firstProposedEvent.smallChanges
+                        let collapsedSmallEvent = ArticleAsLivingDocViewModel.TypedEvent.small(.init(smallChanges: smallChanges))
+
+                        let proposedSectionCollapsed = proposedSection
+                        let currentSectionCollapsed = currentSection
+
+                        proposedSectionCollapsed.typedEvents.removeFirst()
+                        currentSectionCollapsed.typedEvents.removeLast()
+
+                        currentSnapshot.deleteItems([lastCurrentEvent])
+                        currentSnapshot.appendItems([collapsedSmallEvent], toSection: currentSection)
+                        currentSnapshot.appendItems(proposedSectionCollapsed.typedEvents, toSection: currentSection)
+                        existingSections.append(proposedSectionCollapsed)
+                    } else {
+                        currentSnapshot.appendItems(proposedSection.typedEvents, toSection: currentSection)
+                        existingSections.append(proposedSection)
+                    }
                 }
             }
         }
 
+        // Collapse first proposed section into last current section if both only contain small events
+        if
+            let lastCurrentSection = currentSnapshot.sectionIdentifiers.last, lastCurrentSection.containsOnlySmallEvents,
+            let firstProposedSection = sections.first, firstProposedSection.containsOnlySmallEvents {
```

Comment:
Should this first proposed section be the first section that doesn't contain the existing sections?

Reply:
I certainly could be missing something, but I _believe_ it's sufficient without checking that in practice. That is, I think `!existingSections.contains(firstProposedSection)` will always evaluate to `true` in that if statement.

Reply:
Sounds reasonable to me - I think when I wrote that I neglected to consider that we're at most combining two sections into one in this block.